Copy Content-Type from Netty HttpResponse to APIGatewayV2HTTPResponse

richiethom opened the pull request:
Fixes #15909 

If the Netty response is an HttpResponse, retrieve its Content-Type and set it on the APIGatewayV2HTTPResponse.

(my first pull request for a long time, a lot has changed, please don't hesitate to fix/squash if necessary)

gsmet commented on extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java:
```diff
@@ -92,6 +94,16 @@ public void handleMessage(Object msg) {
+                        final String contentType = responseHeaders.getAsString(HttpHeaderNames.CONTENT_TYPE);
+                        if (contentType != null && contentType.trim().length() > 0) {
+                            final Map<String, String> headers = new HashMap<>();
+                            headers.put("Content-Type", contentType);
```
Shouldn't you use the constant here too?

Also could you squash the commits? Thanks!

richiethom replied:
I wasn't sure about using the constant there too. The constant is an `AsciiString`, not a `String` so I need to call `toString()` on it. In addition, the constant is from Netty, whereas the `Map` is being put into AWS. Very unlikely I know, but if the Netty headers started being named differently, in a Netty-specific way, it might not work properly on AWS.

If you still want me to use the constant, I can:

`headers.put(HttpHeaderNames.CONTENT_TYPE.toString(), contentType);`

Once you've confirmed one way or the other, I'll work out how to squash the commits!

gsmet replied:
Yeah, keep it as it is, I totally agree with you it's better this way.

richiethom committed:
Replace MultiValueHeaders with normal Headers to match AWS APIGateway V2

richiethom commented:
@gsmet finally worked out how to do squash the PR, hopefully it's all ok now

richiethom commented on extensions/amazon-lambda-http/runtime/src/main/java/io/quarkus/amazon/lambda/http/LambdaHttpHandler.java:
```diff
@@ -92,6 +94,16 @@ public void handleMessage(Object msg) {
+
                     Headers multiValueHeaders = new Headers();
                     responseBuilder.setMultiValueHeaders(multiValueHeaders);
                     for (String name : res.headers().names()) {
```
I saw that, but without my change the Lambda is returning 'text/plain' rather than 'application/json'. Digging a bit further I found ["Format 2.0 doesn't have multiValueHeaders or multiValueQueryStringParameters fields"](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html) on the AWS API Gateway documentation.

(as stated in #15909 I'm using API Gateway v2)

If you'd rather I used your approach to copy across all headers, but use the setHeaders() method instead, I can easily do this.

patriot1burke replied:
Richie,

The code uses com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPResponse which does have a multivalueheaders property.   I guess when they run within Lambda runtime without our wrapper they marshal this correctly?  Oh well, your fix looks good.

richiethom commented:
I've changed it all, the MultiValueHeaders stuff is now replaced with 'normal' headers. If the Header returned from Netty is multi-valued, it is concatenated with commas, as explained in the [AWS API Gateway 2 documentation](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html):

> Format 2.0 doesn't have multiValueHeaders or multiValueQueryStringParameters fields. Duplicate headers are combined with commas and included in the headers field. Duplicate query strings are combined with commas and included in the queryStringParameters field.

Let me know what you think.

richiethom commented:
hi @patriot1burke and @gsmet this is now ready for a review.

patriot1burke commented:
@richiethom Great job on this.  Thank you.
